File: editor/src/main/java/com/libreshockwave/editor/score/ScoreModel.java
package com.libreshockwave.editor.score;

import java.awt.*;

/**
 * Data model bridging ScoreChunk data to the ScorePanel UI.
 * Provides frame count, channel count, and cell data for rendering.
 */
public class ScoreModel {

    private int frameCount;
    private int channelCount;
    // Will be populated from ScoreChunk data
    private Color[][] cellColors;

    public ScoreModel() {
        this.frameCount = 0;
        this.channelCount = 0;
    }

    public ScoreModel(int frameCount, int channelCount) {
        this.frameCount = frameCount;
        this.channelCount = channelCount;
        this.cellColors = new Color[channelCount][frameCount];
    }

    public int getFrameCount() {
        return frameCount;
    }

    public int getChannelCount() {
        return channelCount;
    }

    public Color getCellColor(int channel, int frame) {
        if (cellColors == null || channel >= channelCount || frame >= frameCount) {
            return null;
        }
        return cellColors[channel][frame];
    }

    public void setCellColor(int channel, int frame, Color color) {
        if (cellColors != null && channel < channelCount && frame < frameCount) {
            cellColors[channel][frame] = color;
        }
    }
}

File: cpp/src/editor/score/ScoreModel.cpp
#include "libreshockwave/editor/score/ScoreModel.hpp"

namespace libreshockwave::editor::score {

ScoreModel::ScoreModel() = default;

ScoreModel::ScoreModel(int frameCount, int channelCount)
    : frameCount_(frameCount),
      channelCount_(channelCount) {
    if (frameCount_ > 0 && channelCount_ > 0) {
        cellColors_.resize(static_cast<std::size_t>(channelCount_));
        for (auto& row : cellColors_) {
            row.resize(static_cast<std::size_t>(frameCount_));
        }
    }
}

int ScoreModel::frameCount() const {
    return frameCount_;
}

int ScoreModel::channelCount() const {
    return channelCount_;
}

std::optional<ScoreColor> ScoreModel::cellColor(int channel, int frame) const {
    if (cellColors_.empty() || channel < 0 || frame < 0 ||
        channel >= channelCount_ || frame >= frameCount_) {
        return std::nullopt;
    }
    return cellColors_[static_cast<std::size_t>(channel)][static_cast<std::size_t>(frame)];
}

void ScoreModel::setCellColor(int channel, int frame, std::optional<ScoreColor> color) {
    if (cellColors_.empty() || channel < 0 || frame < 0 ||
        channel >= channelCount_ || frame >= frameCount_) {
        return;
    }
    cellColors_[static_cast<std::size_t>(channel)][static_cast<std::size_t>(frame)] = color;
}

} // namespace libreshockwave::editor::score
